apps/examples/grpc_route_client/route_guide_client.cc

```diff
@@ -252,40 +270,292 @@ class RouteGuideClient {
+    Status status;
+    ClientContext context;
+
+    rect.mutable_lo()->set_latitude(400000000);
+    rect.mutable_lo()->set_longitude(-750000000);
+    rect.mutable_hi()->set_latitude(420000000);
+    rect.mutable_hi()->set_longitude(-730000000);
+    std::cout << "Looking for features between 40, -75 and 42, -73"
+              << std::endl;
+
+    std::unique_ptr<ClientAsyncReader<Feature> > reader(
+       stub_->AsyncListFeatures(&context, rect, &cq, (void*)1));
+
+    void* got_tag;
+    bool ok = false;
+    bool ret;
+
+    ret = cq.Next(&got_tag, &ok);
+    if (!ret || !ok || got_tag != (void*)1) {
+      std::cout << "RPC initialization Failed" << std::endl;
+      return;
+    }
+
+    while (1) {
+      reader->Read(&feature, got_tag);
+      ok = false;
+      bool ret = cq.Next(&got_tag, &ok);
+      if (!ret || !ok || got_tag != (void*)1) {
+        break;
+      }
+      std::cout << "Found feature called "
+                << feature.name() << " at "
+                << feature.location().latitude()/kCoordFactor_ << ", "
+                << feature.location().longitude()/kCoordFactor_ << std::endl;
+    }
+    reader->Finish(&status, (void*)1);
+    ret = cq.Next(&got_tag, &ok);
+    if (!ret || !ok || got_tag != (void*)1) {
+      std::cout << "reader Finish is Failed" << std::endl;
+      return;
+    }
+
+    if(status.ok()) {
+       std::cout << "ListFeature Done" << std::endl;
+    } else {
+       std::cout << "ListFeature Failed" << std::endl;
+    }
+}
+
+  void RecordRoute() {
+    Point point;
+    RouteSummary stats;
+    ClientContext context;
+    CompletionQueue cq;
+    const int kPoints = 10;
+    Status status;
+    unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
+
+    std::default_random_engine generator(seed);
+    std::uniform_int_distribution<int> feature_distribution(
+        0, feature_list_.size() - 1);
+    std::uniform_int_distribution<int> delay_distribution(
+        500, 1500);
+
+    std::unique_ptr<ClientAsyncWriter<Point>> writer(
+        stub_->AsyncRecordRoute(&context, &stats, &cq, (void*)1));
+
+    void* got_tag;
+    bool ok = false;
+    bool ret;
+
+    ret = cq.Next(&got_tag, &ok);
+    if (!ret || !ok || got_tag != (void*)1) {
+      std::cout << "RPC initialization Failed" << std::endl;
+      return;
+    }
+
+    for (int i = 0; i < kPoints; i++) {
+      const Feature& f = feature_list_[feature_distribution(generator)];
+      std::cout << "Visiting point "
+                << f.location().latitude()/kCoordFactor_ << ", "
+                << f.location().longitude()/kCoordFactor_ << std::endl;
+      writer->Write(f.location(), got_tag);
+      ret = cq.Next(&got_tag, &ok);
+      if (!ret || !ok || got_tag != (void*)1) {
+        std::cout << "Write Failed" << std::endl;
+        break;
+      }
+      std::this_thread::sleep_for(std::chrono::milliseconds(
+        delay_distribution(generator)));
+    }
+
+    writer->WritesDone(got_tag);
```

Comment:
Is WritesDone necessary to be called? Doxygen says that Finish method can implicitly verify that client has sent all messages. Could you please verify?

Reply:
yes, WritesDone is nessasary for verify to finish Writes on this stream.
and then it is actually done when its CQ tag comes back.

Reply:
Thank you for verifying. Yes, WritesDoneis necessary. Please ignore the comment above.